core/chains/terra/terratxm/txm.go

```diff
@@ -157,23 +157,45 @@ func unmarshalMsg(msgType string, raw []byte) (sdk.Msg, string, error) {
 	return nil, "", errors.Errorf("unrecognized message type: %s", msgType)
 }
 
+type msgValidator struct {
+	cutoff         time.Time
+	expired, valid terra.Msgs
+}
+
+func (e *msgValidator) add(msg terra.Msg) {
+	if msg.CreatedAt.Before(e.cutoff) {
+		e.expired = append(e.expired, msg)
+	} else {
+		e.valid = append(e.valid, msg)
+	}
+}
+
 func (txm *Txm) sendMsgBatch(ctx context.Context) {
-	var notExpired, expired terra.Msgs
+	msgs := msgValidator{cutoff: time.Now().Add(-txm.cfg.TxMsgTimeout())}
 	err := txm.orm.q.Transaction(func(tx pg.Queryer) error {
-		unstarted, err := txm.orm.GetMsgsState(db.Unstarted, txm.cfg.MaxMsgsPerBatch(), pg.WithQueryer(tx))
+		started, err := txm.orm.GetMsgsState(db.Started, txm.cfg.MaxMsgsPerBatch(), pg.WithQueryer(tx))
+		if err != nil {
+			txm.lggr.Errorw("unable to read unstarted msgs", "err", err)
+			return err
+		}
+		for _, msg := range started {
+			msgs.add(msg)
+		}
+		unstarted, err := txm.orm.GetMsgsState(db.Unstarted, txm.cfg.MaxMsgsPerBatch()-int64(len(started)), pg.WithQueryer(tx))
 		if err != nil {
 			txm.lggr.Errorw("unable to read unstarted msgs", "err", err)
 			return err
 		}
-		cutoff := time.Now().Add(-txm.cfg.TxMsgTimeout())
 		for _, msg := range unstarted {
-			if msg.CreatedAt.Before(cutoff) {
-				expired = append(expired, msg)
-			} else {
-				notExpired = append(notExpired, msg)
-			}
+			msgs.add(msg)
+		}
+		err = txm.orm.UpdateMsgs(unstarted.GetIDs(), db.Started, nil, pg.WithQueryer(tx))
```

Comment:
don't we want to mark only the unstarted and _also not expired_ ones as started here?

Reply:
Yeah, will update.